runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/UnboundedReadFromBoundedSource.java

```diff
@@ -288,6 +288,15 @@ private void init(
             residualElementsList == null
                 ? new ResidualElements(Collections.emptyList())
                 : new ResidualElements(residualElementsList);
+
+        if (this.residualSource != null) {
```

Comment:
Just checking - this means that we are calling `init()` again without ever calling `close()` on the reader?

Reply:
Yes, it's at [Reader.getCheckpointMark()](
https://github.com/apache/beam/blob/204c69721a8d5ef98423ff03bea0b6ab90b10dc3/runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/UnboundedReadFromBoundedSource.java#L385). 

The `getCheckpointMark()` is called at bundle finish in Dataflow streaming jobs, and `close()` will be pending as readers are cached. A reader can be reused in next bundles.